Fix race conditions in NetworkPolicyController

tnqn opened the pull request:
There were a few race conditions in NetworkPolicyController:
* An AppliedToGroup or AddressGroup in use may be removed if situations
like this happens:
1. addANP creates a group for ANP A;
2. addNetworkPolicy reuses the group for KNP B, is going to create an
   internal NetworkPolicy;
3. deleteANP deletes the group for ANP A because at that moment no other
   internal NetworkPolicies are using the group;
4. addNetworkPolicy commits the internal NetworkPolicy for KNP B to
   storage, but the group no longer exists.

* An Antrea-native NetworkPolicy may be out-of-date if situations like
this happens:
1. An ACNP event is received, `updateCNP` calculates the new internal
   NetworkPolicy for the ACNP, is going to commit it to storage;
2. A ClusterGroup event triggers update of the ACNP via
   triggerCNPUpdates
3. triggerCNPUpdates calls reprocessCNP which updates the new internal
   NetworkPolicy for the ACNP and commits it to storage;
4. updateCNP in the first step commits its internal NetworkPolicy to
   storage which overrides the update of the ClusterGroup event.

The second one caused test flake of the test case
"TestGroupNoK8sNP/Case=ACNPNestedClusterGroup".

To resolve the race conditions completely and make NetworkPolicy
handling less error prone, this patch refactors NetworkPolicyController:
* Event handlers no longer update the storage of internal NetworkPolicy
  directly and only triggers resync of affected policies, which ensures
  that there is at most one worker handling an internal NetworkPolicy at
  any moment.
* Ensure atomicity when updating internal NetworkPolicy and creating or
  deleting AddressGroups and AppliedToGroups.

Duplicate code and tests are deleted with the refactoring.

Signed-off-by: Quan Tian <qtian@vmware.com>

tnqn commented:
/test-all

antoninbas commented on pkg/controller/networkpolicy/clusternetworkpolicy.go:
```diff
@@ -218,15 +102,11 @@ func (n *NetworkPolicyController) updateNamespace(oldObj, curObj interface{}) {
+	if labels.Equals(oldNamespace.Labels, curNamespace.Labels) {
+		return
 	}
+	n.enqueueCNPsWithPerNamespaceRule()
```
I was trying to understand why `PerNamespaceSelectors` and all the related stuff around it is not needed anymore. Is it just because we are simplifying the logic and instead of enqueuing a smaller set of CNPs based on which Namespace labels have changed, we enqueue all CNPs with per-Namespace rules? Is it something worth mentioning in the commit message / PR description?

tnqn replied:
Yes, we cannot use per-Namespace index of internalNetworkPolicy because there could be a race condition:
1. AddCNP triggers syncInternalNetworkPolicy
2. syncInternalNetworkPolicy calls getAffectedNamespacesForAppliedTo, in which namespaces that match selfNamespace selectors are retrieved, it doesn't commit the internalNetworkPolicy to storage yet
3. AddNamespace doesn't find the above internalNetworkPolicy, so it doesn't trigger syncInternalNetworkPolicy
4. following step 2, syncInternalNetworkPolicy commits the internalNetworkPolicy to storage, it will be in a stale state

We should use per-Namespace index of ClusterNetworkPolicy, in same situation:
1. AddCNP triggers syncInternalNetworkPolicy
2. syncInternalNetworkPolicy calls getAffectedNamespacesForAppliedTo, in which namespaces that match selfNamespace selectors are retrieved, it doesn't commit the internalNetworkPolicy to storage yet
3. AddNamespace calls finds the ClusterNetworkPolicy, it trigger syncInternalNetworkPolicy
4. following step 2, syncInternalNetworkPolicy commits the internalNetworkPolicy to storage, it will be in a stale state
5. Because step 3 triggerred another syncInternalNetworkPolicy, this time it will count the new Namespace in.

I thought there shouldn't many per-Namespace rules when making the above change, but I agree it's probably an irrelavant change, I will continue filtering a smaller set in this PR.

antoninbas commented on pkg/controller/networkpolicy/clusternetworkpolicy_test.go:
```diff
@@ -1405,739 +1477,62 @@ func TestProcessClusterNetworkPolicy(t *testing.T) {
+	_, npc := newController()
+	cnp := getCNP()
+	newCNP := cnp.DeepCopy()
+	newCNP.Spec.Priority = float64(100)
```
could you add a comment explaining why we set the priority to this value here?

tnqn replied:
added

antoninbas commented on pkg/controller/networkpolicy/antreanetworkpolicy_test.go:
```diff
@@ -583,173 +582,38 @@ func TestProcessAntreaNetworkPolicy(t *testing.T) {
+	_, npc := newController()
+	anp := getANP()
+	newANP := anp.DeepCopy()
+	newANP.Spec.Priority = float64(1)
```
could you add a comment explaining why we set the priority to this value here?

antoninbas commented on pkg/controller/networkpolicy/crd_utils_test.go:
```diff
@@ -428,7 +428,7 @@ func TestToAntreaPeerForCRD(t *testing.T) {
 			npc.addClusterGroup(&cgA)
 			npc.cgStore.Add(&cgA)
-			actualPeer := npc.toAntreaPeerForCRD(tt.inPeers, testCNPObj, tt.direction, tt.namedPortExists)
+			actualPeer := npc.toAntreaPeerForCRD(getAntreaClusterNetworkPolicyReference(testCNPObj), tt.inPeers, testCNPObj, tt.direction, tt.namedPortExists)
```
I just thought about this: could we use `getACNPReference` / `getANPReference` as function names? It gets pretty long to read and we already use the ACNP / ANP abbreviations in other function names

tnqn replied:
updated

antoninbas commented on pkg/controller/networkpolicy/networkpolicy_controller.go:
```diff
@@ -204,9 +222,16 @@ type NetworkPolicyController struct {
+	// addressGroupReferences tracks the reference count of policies for AddressGroups.
+	addressGroupReferences map[string]networkPolicyReferences
+	// appliedToGroupMutex prevents race conditions between multiple internalNetworkPolicyWorkers when they create or
+	// delete same AppliedToGroups and ensures atomicity of updating appliedToGroupStore and appliedToGroupReferences.
```
delete the same

tnqn replied:
fixed

antoninbas commented on pkg/controller/networkpolicy/networkpolicy_controller.go:
```diff
@@ -204,9 +222,16 @@ type NetworkPolicyController struct {
+	// delete same AppliedToGroups and ensures atomicity of updating appliedToGroupStore and appliedToGroupReferences.
+	appliedToGroupMutex sync.Mutex
+	// addressGroupMutex prevents race conditions between multiple internalNetworkPolicyWorkers when they create or
+	// delete same  AddressGroups and ensures atomicity of updating addressGroupStore and addressGroupReferences.
```
delete the same

there is also an extra whitespace

tnqn replied:
fixed

antoninbas reviewed (commented):
I am not super familiar with this code at this stage, but I took a stab at a review

GraysonWu commented on pkg/controller/networkpolicy/networkpolicy_controller.go:
```diff
@@ -1340,77 +1339,130 @@ func (n *NetworkPolicyController) getClusterGroupWorkloads(group *antreatypes.Gr
-		// AddressGroups.
-		klog.V(4).Infof("Internal NetworkPolicy %s Node span remains unchanged. No need to enqueue AddressGroups.", key)
-		return nil
+		utilsets.MergeString(newNodeNames, appGroup.NodeNames)
```
Maybe not related to this change, just come up with this question while reviewing.
Is there a chance that when we read the `appGroup.NodeNames` here, the `NodeNames` hasn't been calculated by `func (n *NetworkPolicyController) syncAppliedToGroup(key string) error`

GraysonWu replied:
NVM. Stupid question.

GraysonWu commented on pkg/controller/networkpolicy/crd_utils.go:
```diff
@@ -177,52 +182,30 @@ func (n *NetworkPolicyController) svcRefToPeerForCRD(svcRefs []v1alpha1.Namespac
+			Name: key,
+		}
+		n.addressGroupStore.Create(addressGroup)
+		klog.V(2).Infof("Created new AddressGroup %v corresponding to ClusterGroup CRD %s", key, intGrp.Name)
```
InfoS

tnqn replied:
fixed

GraysonWu commented on pkg/controller/networkpolicy/networkpolicy_controller.go:
```diff
@@ -1340,77 +1339,130 @@ func (n *NetworkPolicyController) getClusterGroupWorkloads(group *antreatypes.Gr
-func (n *NetworkPolicyController) syncInternalNetworkPolicy(key string) error {
+func (n *NetworkPolicyController) syncInternalNetworkPolicy(key controlplane.NetworkPolicyReference) error {
+	internalNetworkPolicyName := getInternalNetworkPolicyName(key)
+	klog.V(2).Infof("Syncing internal NetworkPolicy %s", key)
```
ditto and other places.

tnqn replied:
actually these are existing logs, they appear as diff somehow because of move of some lines.
I would try to change them to structured logging since they are pointed out, but I may not do further log cleanup as it's not the purpose of the PR and would increase the burdon of review given it already adds 1000+ lines and removes 2000+ lines.

GraysonWu replied:
Make sense.

GraysonWu commented on pkg/controller/networkpolicy/endpoint_querier_test.go:
```diff
@@ -156,6 +157,9 @@ func makeControllerAndEndpointQuerier(objects ...runtime.Object) *endpointQuerie
 	querier := NewEndpointQuerier(c.NetworkPolicyController)
 	// start informers and run controller
 	c.informerFactory.Start(stopCh)
+	c.crdInformerFactory.Start(stopCh)
+	go c.groupingController.Run(stopCh)
+	go c.groupingInterface.Run(stopCh)
```
Why we don't have these before and add them now?

tnqn replied:
A complete workflow needs to start crdInformerFactory, groupingController, groupingInterface, otherwise workers wouldn't run.
Previously it can work because event handlers could also create internalNetworkPolicies and groups, now that event handlers only trigger workers, we need to ensure workers running.

GraysonWu replied:
Got it. Thanks.

tnqn commented:
/skip-integration tested manually

qiyueyao commented:
Regarding PR #4047, has updated documentation in the `enableLogging` section that users should avoid annotation update to prevent NP from working in a stale state. The comment needs to be deleted after this PR fixes race conditions. Thanks!

tnqn commented on pkg/controller/networkpolicy/antreanetworkpolicy.go:
```diff
@@ -132,7 +88,7 @@ func (n *NetworkPolicyController) processAntreaNetworkPolicy(np *crdv1alpha1.Net
 	for _, at := range np.Spec.AppliedTo {
 		appliedToGroupNamesSet.Insert(n.createAppliedToGroup(
-			np.Namespace, at.PodSelector, at.NamespaceSelector, at.ExternalEntitySelector))
+			networkPolicyRef, np.Namespace, at.PodSelector, at.NamespaceSelector, at.ExternalEntitySelector))
```
made some changes to how to track references. Now only policy UID is required for reference, so namespace is still passed.

tnqn replied:
changed to ensure atomicity when updating internal NetworkPolicy and creating or deleting AddressGroups and AppliedToGroups, no extra argument is needed.

tnqn commented on pkg/controller/networkpolicy/clustergroup.go:
```diff
@@ -263,7 +264,19 @@ func (c *NetworkPolicyController) syncInternalGroup(key string) error {
-func (c *NetworkPolicyController) triggerParentGroupSync(grp string) {
+// triggerDerivedGroupUpdates triggers processing of AppliedToGroup and AddressGroup derived from the provided ClusterGroup.
+func (c *NetworkPolicyController) triggerDerivedGroupUpdates(cg string) {
+	_, exists, _ := c.appliedToGroupStore.Get(cg)
+	if exists {
+		c.enqueueAppliedToGroup(cg)
```
added

tnqn commented on pkg/controller/networkpolicy/clustergroup_test.go:
```diff
@@ -883,23 +883,13 @@ func TestSyncInternalGroup(t *testing.T) {
-	// cnp2 is added after the ClusterGroup.
-	npc.addCNP(cnp2)
-
+	// cnp1 is added before the ClusterGroup. The rule's From should be empty as the ClusterGroup hasn't been synced,
```
fixed

tnqn commented on pkg/controller/networkpolicy/clustergroup_test.go:
```diff
@@ -919,11 +908,32 @@ func TestSyncInternalGroup(t *testing.T) {
+	require.NoError(t, err)
+	require.Equal(t, 2, npc.internalNetworkPolicyQueue.Len())
+	for i := 0; i < 2; i++ {
+		key, _ := npc.internalNetworkPolicyQueue.Get()
+		npc.internalNetworkPolicyQueue.Done(key)
```
added

tnqn commented on pkg/controller/networkpolicy/antreanetworkpolicy.go:
```diff
@@ -24,19 +24,22 @@ import (
 	antreatypes "antrea.io/antrea/pkg/controller/types"
 )
 
+func getAntreaNetworkPolicyReference(anp *crdv1alpha1.NetworkPolicy) controlplane.NetworkPolicyReference {
```
`internalNetworkPolicyQueue` must use value instead of pointer as the key, otherwise the same NetworkPolicies will not be treated as same item because the pointers may be different. But other places could use pointers to save some copy overhead. I have updated them.

tnqn commented:
/test-all

tnqn commented on pkg/controller/networkpolicy/clustergroup.go:
```diff
@@ -261,31 +261,38 @@ func getClusterGroupSourceRef(cg *crdv1alpha3.ClusterGroup) *controlplane.GroupR
 }
 
+// triggerDerivedGroupUpdates triggers processing of AppliedToGroup and AddressGroup derived from the provided group.
+func (c *NetworkPolicyController) triggerDerivedGroupUpdates(grp string) {
```
This replaces the previous group enqueue operations in `reprocessCNP`, which blindly enqueues all AppliedToGroups and AddressGroups.

tnqn commented:
/test-all

tnqn commented:
/test-all
/test-ipv6-all
/test-ipv6-only-all
/test-windows-all

tnqn commented:
/test-all

antoninbas commented on pkg/controller/networkpolicy/networkpolicy_controller.go:
```diff
@@ -908,66 +859,17 @@ func (n *NetworkPolicyController) enqueueAppliedToGroup(key string) {
-	n.internalNetworkPolicyQueue.Add(key)
+func (n *NetworkPolicyController) enqueueInternalNetworkPolicy(key *controlplane.NetworkPolicyReference) {
+	klog.V(4).Infof("Adding new key %v to internal NetworkPolicy queue", key)
+	// It must use value instead of pointer as the key, otherwise the same NetworkPolicies will not be treated as same
+	// item because the pointers may be different.
```
thanks for adding this comment

antoninbas commented on pkg/controller/networkpolicy/antreanetworkpolicy.go:
```diff
@@ -104,33 +72,20 @@ func (n *NetworkPolicyController) deleteANP(old interface{}) {
-// in case of ADD event or modified and store the updated instance, in case
-// of an UPDATE event.
-func (n *NetworkPolicyController) processAntreaNetworkPolicy(np *crdv1alpha1.NetworkPolicy) *antreatypes.NetworkPolicy {
+// instance to the caller wherein.
```
remove "wherein", which is no longer valid

tnqn replied:
removed

antoninbas reviewed (dismissed):
LGTM< I think @GraysonWu should take another look as well

GraysonWu reviewed (dismissed):
LGTM. Just curious: is there a way to "test" a race condition?

tnqn commented:
> LGTM. Just curious: is there a way to "test" a race condition?

Good question, I added an unit test `TestSyncInternalNetworkPolicyConcurrently` to verify one of the race conditions is resolved. Without the lock `n.internalNetworkPolicyMutex.Lock()`, the test had a chance to fail:
```
# go test -run TestSyncInternalNetworkPolicyConcurrently  -count 500 -race ./pkg/controller/networkpolicy/
--- FAIL: TestSyncInternalNetworkPolicyConcurrently (0.01s)
    networkpolicy_controller_test.go:3155:
                Error Trace:    networkpolicy_controller_test.go:3155
                                                        networkpolicy_controller_test.go:3130
                Error:          "[]" should have 1 item(s), but has 0
                Test:           TestSyncInternalNetworkPolicyConcurrently
    networkpolicy_controller_test.go:3158:
                Error Trace:    networkpolicy_controller_test.go:3158
                                                        networkpolicy_controller_test.go:3130
                Error:          Should be true
                Test:           TestSyncInternalNetworkPolicyConcurrently
FAIL
FAIL    antrea.io/antrea/pkg/controller/networkpolicy   5.405s
FAIL
```
Although the code before refactoring didn't create and delete groups in `SyncInternalNetworkPolicy` but in `addNetworkPolicy`, `addANP`, `addCNP`, `updateNetworkPolicy`, etc., they could execute concurrently as they are event handlers of different resources, so the same race condition could happen like above.

tnqn commented:
/skip-all the latest change only added an unit test and updated a comment.

tnqn commented:
@antoninbas @GraysonWu could you give another approval?

Dyanngg commented on pkg/controller/networkpolicy/antreanetworkpolicy.go:
```diff
@@ -249,26 +210,23 @@ func (e ErrNetworkPolicyAppliedToUnsupportedGroup) Error() string {
-	// Find the internal Group corresponding to this Group
+func (n *NetworkPolicyController) createAppliedToGroupForNamespacedGroup(namespace, groupName string) (*antreatypes.AppliedToGroup, error) {
+	// Namespaced group uses NAMESPACE/NAME as the key of the corresponding internal group.
+	key := k8s.NamespacedName(namespace, groupName)
```
I'm not sure if the name `internalGroupKeyFunc` still make sense since it is only applicable to ClusterGroup now.:
```
// internalGroupKeyFunc knows how to generate the key for an internal Group based on the object metadata
// of the corresponding ClusterGroup resource. Currently the Name of the ClusterGroup is used to ensure uniqueness.
func internalGroupKeyFunc(obj metav1.Object) string {
	return obj.GetName()
}
```
For better readability, we can either 1. use a type assertion and return different key for Namespaced Group and ClusterGroup 2. have separate key funcs for these resources

tnqn replied:
It has been changed by https://github.com/antrea-io/antrea/pull/2438
https://github.com/antrea-io/antrea/blob/65b62cc645e83cce0a605a40422d307bf1767d4a/pkg/controller/networkpolicy/networkpolicy_controller.go#L1556-L1561

tnqn commented on pkg/controller/networkpolicy/antreanetworkpolicy.go:
```diff
@@ -249,26 +210,23 @@ func (e ErrNetworkPolicyAppliedToUnsupportedGroup) Error() string {
-		return "", nil
+		// Internal Group is not found, which means the corresponding namespaced group is either not created yet or not
+		// processed yet. Once the internal Group is created and processed, the sync worker for internal group will
+		// re-enqueue the ClusterNetworkPolicy processing which will trigger the creation of AppliedToGroup.
```
thanks for pointing it out, will fix it in other PR

tnqn commented:
/skip-all

jianjuns commented:
@tnqn Sorry for the late review. I added several questions about locking in the comments, and have a general question: why earlier we chose to process related NPs in event handlers directly? Any side effects (performance, etc.) with the new approach of enqueuing NPs?

tnqn commented:
> @tnqn Sorry for the late review. I added several questions about locking in the comments

@jianjuns I don't find your questions about locking in the comments. Have you published them?

> and have a general question: why earlier we chose to process related NPs in event handlers directly? 

It's because we only support K8s NetworkPolicy when designing the workflow, it worked fine and was more efficient to calculate spec in event handlers and calculate span in workers separately. But with Antrea-native policies, groups features added, the workflow lost its advantage and became complex and error prone:

When we only support K8s NetworkPolicy:
1. InternalNetworkPolicy, AddressGroup, AppliedToGroup's spec have only single source of truth, which is K8s NetworkPolicy. Only K8s NetworkPolicy creation, update, and deletion events could trigger their creation and deletion, and their spec's update.
2. Event handlers of a single kind of resources (e.g. addNetworkPolicy, updateNetworkPolicy, deleteNetworkPolicy) are executed sequentially, so creating and deleting InternalNetworkPolicies, AddressGroups, AppliedToGroups don't need to concern race conditions.

The above two facts changed after NetworkPolicyController started to support Antrea-native policies and groups:

1. InternalNetworkPolicy, AddressGroup, AppliedToGroup's spec have multiple source of truth: K8s NetworkPolicy, Antrea ClusterNetworkPolicy, Antrea NetworkPolicy, Group, ClusterGroup, Namespace's annotation (for determining Audit Logging property of K8s NetworkPolicy). Then all these resources' events can trigger InternalNetworkPolicy, AddressGroup, and AppliedToGroup's creation, deletion, and spec's update. That's why there are functions like triggerANPUpdates, triggerCNPUpdates, triggerParentGroupSync, etc.
2. Event handlers of different kind of resources can execute concurrently, so race conditions between them need to be taken into consideration. Also, as ClusterGroup, Group, and even Namespace's annotation can affect the spec of InternalNetworkPolicy, the workers that sync ClusterGroup, Group and the event handlers of Namespace can update InternalNetworkPolicy, which can also conflict with the event handlers of networkpolicies.

Mutex need to be used very carefully to avoid race conditions, for example, even just internal NetworkPolicy, they may be written by the following goroutines:

1. K8s NetworkPolicy event handler executor
2. Antrea ClusterNetworkPolicy event handler executor
3. Antrea NetworkPolicy event handler executor
4. Namespace event handler executor
5. multiple InternalNetworkPolicy workers
6. multiple InternalGroup workers

But even mutex is used properly, there will be a bottleneck for all the above executors, only one of them could execute the code block that may create/update/delete InternalNetworkPolicy, AddressGroup, AppliedToGroup. The previous code's critical section was not big enough to avoid race conditions, hence the errors. To fix it in old way, more code block need to be executed with lock acquired. And the code was quite redundant, this can be told by the lines of code change: the refactoring added 1,504 lines of code and deleted 2,735 lines of code, and the unit test coverage of this package was still improved from 58.2% to 63.3%.

I just created issue #4127 to describe the race conditions and why a refactoring is necessary for record.

> Any side effects (performance, etc.) with the new approach of enqueuing NPs?

In theory, `syncInternalNetworkPolicy`'s overhead is greater than before because it calculates both spec and span while it only calculates span before. But the concurrency is improved, previously their spec calculation is executed sequentially, and the use of `internalNetworkPolicyMutex` also makes span calculation is executed sequentially even though there are multiple workers, now spec and span calculation are executed concurrently and can scale out by increasing number of workers, only updating storage is executed sequentially. In theory, calculating spec should not cost much as it's a simple translation operation, unlike calculating span operation which may depend on the number of Nodes, so I think repeating calculating spec in `syncInternalNetworkPolicy` should be fine. Even it adds some overhead for each round of processing, it should still be better than serializing them in the above 6 executors.

jianjuns reviewed (commented):
Thanks Quan for the detailed explanation. All make sense to me.

Yes seems I forgot to publish the previous comments.

tnqn commented on pkg/controller/networkpolicy/networkpolicy_controller.go:
```diff
@@ -1447,76 +1348,182 @@ func (n *NetworkPolicyController) getInternalGroupWorkloads(group *antreatypes.G
+		n.internalNetworkPolicyMutex.Lock()
+		defer n.internalNetworkPolicyMutex.Unlock()
+
+		if !oldInternalPolicyExists {
```
Yes, they need to be protected by the same mutex because NP store has index of AddressGroups and AppliedToGroups, which is used to determine whether an AddressGroup/AppliedToGroup is orphan or not. For example, the following code means the AppliedToGroup is no longer used and can be deleted:
```
objs, _ := n.internalNetworkPolicyStore.GetByIndex(store.AppliedToGroupIndex, atgName)
if len(objs) == 0 {
``` 

We must ensure the AddressGroup/AppliedToGroup data (stored in their own store) and their index (stored in NP store) are updated atomically, otherwise the first race condition the PR description describes could happen:

1. addANP creates a group for ANP A;
2. addNetworkPolicy reuses the group for KNP B, is going to create an internal NetworkPolicy;
3. deleteANP deletes the group for ANP A because at that moment no other internal NetworkPolicies are using the group;
4. addNetworkPolicy commits the internal NetworkPolicy for KNP B to storage, but the group no longer exists.

jianjuns replied:
Ok. Got it.

tnqn commented on pkg/controller/networkpolicy/networkpolicy_controller.go:
```diff
@@ -1447,76 +1348,182 @@ func (n *NetworkPolicyController) getInternalGroupWorkloads(group *antreatypes.G
+	case controlplane.AntreaClusterNetworkPolicy:
+		cnp, err := n.cnpLister.Get(key.Name)
+		if err != nil {
+			n.deleteInternalNetworkPolicy(internalNetworkPolicyName)
```
It's protected by the mutex:
https://github.com/antrea-io/antrea/blob/9c9b098879162903a1250e8fd85c0fa29bb31362/pkg/controller/networkpolicy/networkpolicy_controller.go#L1540-L1542

jianjuns replied:
Humm.. not sure why missed these lines.
